extension/wallet/src/protocol/repay_loan.rs

```diff
@@ -0,0 +1,150 @@
+use baru::loan::Borrower1;
+use elements::{bitcoin::util::amount::Amount, secp256k1_zkp::SECP256K1, Txid};
+use rand::thread_rng;
+use std::str::FromStr;
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+pub async fn repay_loan(wallet_name: String, loan_txid: String) -> Result<JsValue, JsValue> {
+    let loan_txid = map_err_from_anyhow!(Txid::from_str(&loan_txid))?;
+    let txid = map_err_from_anyhow!(repay_loan_w(wallet_name, loan_txid).await)?;
+    let txid = map_err_from_anyhow!(JsValue::from_serde(&txid))?;
+    Ok(txid)
+}
+
+// TODO: Parts of the implementation are very similar to what we do in
+// `sign_and_send_swap_transaction`. We could extract common
+// functionality into crate-local functions
+async fn repay_loan_w(wallet_name: String, loan_txid: Txid) -> Result<Txid, Error> {
+    let esplora_url = {
+        let guard = ESPLORA_API_URL.lock().expect_throw("can get lock");
+        guard.clone()
+    };
+    let client = EsploraClient::new(esplora_url.clone());
+
+    // TODO: Only abort early if this fails because the transaction hasn't been mined
+    if fetch_transaction(esplora_url, loan_txid).await.is_err() {
+        return Err(Error::NoLoan);
+    }
+
+    let storage = Storage::local_storage().map_err(Error::Storage)?;
+
+    let borrower = storage
+        .get_item::<String>(&format!("loan_state:{}", loan_txid))
+        .map_err(Error::Load)?
+        .ok_or(Error::EmptyState)?;
+    let borrower1 = serde_json::from_str::<Borrower1>(&borrower).map_err(Error::Deserialize)?;
+
+    // We are selecting coins with an asset which cannot be
+    // used to pay for fees
+    let zero_fee_rate = 0f32;
+    let zero_fee_offset = Amount::ZERO;
+
+    let wallet_named_cloned = wallet_name.clone();
+    let coin_selector = |amount, asset| async move {
+        let esplora_url = {
+            let guard = ESPLORA_API_URL.lock().expect_throw("can get lock");
+            guard.clone()
+        };
+        let client = EsploraClient::new(esplora_url.clone());
+        let mut wallet = current(&wallet_named_cloned, &LOADED_WALLET)
+            .await
+            .map_err(Error::LoadWallet)?;
+        wallet
+            .sync(&client.clone())
+            .await
+            .map_err(Error::SyncWallet)?;
+        let inputs = wallet.coin_selection(amount, asset, zero_fee_rate, zero_fee_offset)?;
+        Ok(inputs)
+    };
+
+    let signer = |transaction| async move {
+        let esplora_url = {
+            let guard = ESPLORA_API_URL.lock().expect_throw("can get lock");
+            guard.clone()
+        };
+        let client = EsploraClient::new(esplora_url.clone());
+        let mut wallet = current(&wallet_name, &LOADED_WALLET)
+            .await
+            .map_err(Error::LoadWallet)?;
+        wallet
+            .sync(&client.clone())
+            .await
+            .map_err(Error::SyncWallet)?;
+        Ok(wallet.sign(transaction))
+    };
+
```

Comment:
`Wallet::sign` and `Wallet::coin_selection` both take `&self`. I think you should be able to get away with:

1. Unlocking the wallet only once.
2. Syncing the wallet only once.
3. Passing simple, _non_-move async blocks into `loan_repayment_transaction` for coin-selection and signer. Because the references to wallet are not mutable, you can have two at the same time.